### PR TITLE
fix(postgresql): Fix PostgreSQLOidDataType failing to read an oid tha…

### LINKED ISSUE #693: PostgreSQLOidDataType should not fail if OID is not a LargeObject
> **Migrated from SourceForge** — [Bugs #437](https://sourceforge.net/p/dbunit/bugs/437/)
> Originally reported by **coiouhkc** on 2021-01-11 14:03:12.547000
> Priority: `5`
> Assigned to **jeffjensen**
> SF status: pending
> Votes: 1

See https://www.postgresql.org/docs/11/datatype-oid.html, OID data type is not necessarily referring to large object, therefore it's totally OK, if there is no LargeObject mapped to the numerical value, e.g. `'table'::regclass`.
Currently PostgreSQLOidDataType would fail with smth. like `org.dbunit.dataset.DataSetException: org.postgresql.util.PSQLException: ERROR: large object 16548 does not exist`

---

**jeffjensen** — 2021-10-30 15:05:41.456000
Thanks for the report.  Please create a merge request with a proposed fix and a test that shows the problem so it stays fixed.

### PR DESCRIPTION
**Pull request by jeffjensen**
…t isn't a large object

A PostgreSQL oid column is a generic object identifier, not necessarily a large object reference (e.g. a table's own oid via 'table'::regclass, the issue's own example). PostgreSQLOidDataType#getSqlValue() previously had a commented-out swallow-everything attempt for this, explicitly noted as wrong since it would hide real errors - so it just let the SQLException propagate and fail the whole read instead.

* Distinguish "not a large object" from every other failure via the driver's own SQLState (42704/undefined_object, exposed as org.postgresql.util.PSQLState.UNDEFINED_OBJECT) instead of a blanket catch; any other SQLException still propagates unchanged.
* Open the large object under a savepoint: PostgreSQL aborts the entire enclosing transaction on any failed command, so rolling back to the savepoint on failure clears the abort without discarding any other work already done in that transaction, keeping the connection usable for the rest of the read.
* Add Mockito-based PostgreSQLOidDataTypeTest coverage: zero-oid, a genuine large object, a 42704 failure, and a non-42704 SQLException still propagating.
* Extend PostgresSQLOidIT with a live-container case inserting two distinct real catalog oids via raw SQL (dbUnit's own setSqlValue() always creates a genuine large object, so it can't produce this scenario itself); reading the second row also proves the connection recovers, since advancing past row 1 requires a real resultSet.next() round-trip.
* Document the new behavior in postgresql.adoc, including a Known Quirks note that a non-large-object oid reads identically to a genuine SQL NULL.

Refs: 693

Claude-Session: https://claude.ai/code/session_01XBjSJ2bgVtwqyxGv2nvwP9

## Summary by Sourcery

Handle PostgreSQL OID values that do not reference large objects without failing the entire read, and ensure transactions recover cleanly after large-object access errors.

Bug Fixes:
- Treat non-large-object OID values as NULL instead of throwing, based on PostgreSQL’s undefined_object SQLState, while still propagating other SQLExceptions.
- Wrap large-object reads in a savepoint so failed large-object access does not abort the surrounding transaction or break subsequent result set processing.

Enhancements:
- Add unit and integration tests covering zero OID, valid large objects, non-large-object OIDs, and other failure modes for PostgreSQLOidDataType.
- Document the revised OID handling semantics and known quirks for PostgreSQL in the PostgreSQL database guide.

Documentation:
- Update postgresql.adoc to describe how OID columns are handled, including that non-large-object OIDs are read as NULL and noted as a known quirk.

Tests:
- Introduce Mockito-based PostgreSQLOidDataTypeTest for fine-grained behavior verification of getSqlValue() under various OID scenarios.
- Extend PostgresSQLOidIT to exercise real catalog OIDs that are not large objects and verify the connection remains usable across multiple rows.

Chores:
- Record the fix for issue 693 in changes.xml, describing the corrected behavior for PostgreSQLOidDataType#getSqlValue().

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved PostgreSQL `oid` handling for missing or invalid large-object references.
  * Preserved transaction usability after failed large-object reads.
  * Continued propagating unexpected database errors.

* **Documentation**
  * Documented PostgreSQL `oid` read/write behavior, error handling, recovery, and known `NULL` value limitations.

### CODE REVIEW

File: src/test/java/org/dbunit/ext/postgresql/PostgreSQLOidDataTypeTest.java [hunk 1]
```diff
@@ -21,22 +21,76 @@
...
+    @Mock
+    private Savepoint savepoint;
+
+    private final PostgreSQLOidDataType type = new PostgreSQLOidDataType();
+
+    private void mockConnectionChain() throws SQLException
+    {
+        when(resultSet.getStatement()).thenReturn(statement);
+        when(statement.getConnection()).thenReturn(connection);
+        when(connection.getAutoCommit()).thenReturn(true);
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/test/java/org/dbunit/ext/postgresql/PostgreSQLOidDataTypeTest.java]: Added `testGetSqlValue_whenAmbientAutoCommitIsAlreadyFalse_leavesItFalseAndStillRecovers`, verifying `setAutoCommit(true)` is never called and the savepoint/rollback recovery still returns null for a non-large-object oid when the connection comes in with auto-commit already disabled. Parameterized `mockConnectionChain(boolean)` per the suggested structure.

Used a different target method though: the savepoint/rollback logic lives in `getSqlValue()`/`readLargeObject()`, not `setSqlValue()` (which only writes - it always creates a brand-new large object and never opens/reads an existing one), so the suggested test body's `type.setSqlValue(oid, statement, 1)` call and `LargeObjectManager#open()`/`createLO()` stubbing didn't match the method this PR actually changed. Fixed in commit fee5174e.

File: src/main/java/org/dbunit/ext/postgresql/PostgreSQLOidDataType.java [hunk 2]
```diff
@@ -58,28 +60,49 @@ public Object getSqlValue(final int column, final ResultSet resultSet)
...
+     * wrapped in a savepoint: on failure, rolling back to it clears the abort without discarding
+     * any other work already done in that transaction.
+     *
+     * @param connection the connection to read from, already in a non-autocommit transaction.
+     * @param lobj the large object API to read the oid through.
+     * @param oid the oid value read from the result set.
+     * @return the large object's bytes, or <code>null</code> if <code>oid</code> is not a large
+     *         object.
+     * @throws SQLException on any failure other than the oid not referring to a large object.
+     */
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/main/java/org/dbunit/ext/postgresql/PostgreSQLOidDataType.java:100]: Added JavaDoc to `getSqlValue()` and fixed `readLargeObject()`'s `@param`/`@return` lines to start with a capital letter and end with a period. Fixed in commit fee5174e.

File: src/main/java/org/dbunit/ext/postgresql/PostgreSQLOidDataType.java [hunk 3]
```diff
@@ -58,28 +60,49 @@ public Object getSqlValue(final int column, final ResultSet resultSet)
...
-        } finally {
-            connection.setAutoCommit(autoCommit);
+        } catch (SQLException ex) {
+            connection.rollback(savepoint);
+            if (PSQLState.UNDEFINED_OBJECT.getState().equals(ex.getSQLState())) {
+                logger.debug("oid {} is not a large object (SQLState={}), returning null.", oid,
+                    ex.getSQLState());
+                return null;
+            }
+            throw ex;
```
[the remark this answers is not in the document]
Comment (jeffjensen) [reply src/main/java/org/dbunit/ext/postgresql/PostgreSQLOidDataType.java:115]: Agreed, and fixed - `readLargeObject()` now has two separate try/catch blocks: the first wraps only `lobj.open()`, where SQLState 42704 returns null and everything else rethrows; the second wraps `size()`/`read()`/`close()` and always rolls back and rethrows unconditionally, since the oid is already proven to be a real large object by that point.

One deliberate deviation from the suggestion: I didn't add an unconditional `finally { obj.close(); }` for the post-open path. Verified against a live PostgreSQL 16 container that calling `close()` immediately after a failed `size()`/`read()` - before rolling back - itself throws `current transaction is aborted, commands ignored until end of transaction block`, which would replace/mask the original, more meaningful exception. The savepoint rollback already reclaims the large object server-side (it undoes the `open()` itself), so an explicit `close()` on the failure path is both unnecessary and actively unsafe; `close()` is now only called on the success path, where it's safe.

Added tests for both new failure paths (`testGetSqlValue_whenOpenSucceedsButSizeFails_rethrowsAndRecoversTransaction`, `testGetSqlValue_whenReadSucceedsButCloseFails_rethrowsAndRecoversTransaction`). Fixed in commit fee5174e.
